app/src/main/java/com/epfl/drawyourpath/database/FireDatabase.kt

```diff
@@ -38,76 +70,175 @@ class FireDatabase : Database() {
         return future
     }
 
-    override fun updateUsername(username: String, userId: String): CompletableFuture<Boolean> {
-        TODO("This will be implmented during next task when I will clean the database organisation")
+    override fun updateUsername(username: String): CompletableFuture<Boolean> {
+        val future = CompletableFuture<Boolean>()
+
+        val userId = getUserId()
+        //obtain the past username form the userId
+        getUsernameFromUserId(userId).thenAccept{ pastUsername ->
+            if(pastUsername == null){
+                future.completeExceptionally(java.lang.Error("Impossible to find the past username !"))
+            }else{
+                //update the link username to userId and the username on the userAccount
+                setUsername(username).thenAccept{ isSetUsername ->
+                    if(isSetUsername){
+                        //remove the past username from the link username/userId
+                        database.child(usernameToUserIdFileName).child(pastUsername).removeValue()
+                            .addOnSuccessListener { future.complete(true) }
+                            .addOnFailureListener { future.completeExceptionally(java.lang.Error("Impossible to remove the past username link !"))}
+                    }else{
+                        future.completeExceptionally(java.lang.Error("Impossible to set this username !"))
+                    }
+                }
+            }
+        }
+        return future
+    }
+
+    override fun setUsername(username: String): CompletableFuture<Boolean> {
+        val future = CompletableFuture<Boolean>()
+
+        val userId = getUserId()
+        //check if the username is available
+        isUsernameAvailable(username).thenAccept{isAvailable ->
```

Comment:
Username availability shouldn't be handled on the client but through Firebase Security Rules.

This way, we can only try to set it without checking its availability first and it will fail due to the security rules.

We could let it like that for now but add a `TODO` for the future.